pkg/assembler/backends/testing/backend.go

```diff
@@ -16,12 +16,43 @@
 package testing
 
 import (
+	"sync"
+
 	"github.com/guacsec/guac/pkg/assembler/backends"
 	"github.com/guacsec/guac/pkg/assembler/graphql/model"
 )
 
+var (
+	index      = indexType{}
+	packages   = pkgTypeMap{}
+	sources    = srcTypeMap{}
+	sourceMaps = srcMaps{}
+)
+
 type DemoCredentials struct{}
 
+// IDs: We have a global ID for all nodes that have references to/from.
+// Since we always ingest data and never remove, we can keep this global and
+// increment it as needed.
+// For fast retrieval, we also keep a map from ID from nodes that have it.
+// IDs are stored as string in graphql even though we ask for integers
+// See https://github.com/99designs/gqlgen/issues/2561
+type nodeID int
+
+type hasID interface {
+	getID() nodeID
+}
+
+type indexType map[nodeID]hasID
+
+// locking to ensure ID is not duplicated
+func (c *demoClient) getNextID() nodeID {
```

Comment:
hmm im wondering since we're exposing the nodeID now, maybe the query Spec can include NodeId as well and we can have an index here which will be an optimization... but this i think can be done later.

Reply:
so we do have the `ID` as part of the query spec and an index on those: `type indexType map[nodeID]hasID
`